src/probnum/utils/linalg/_orthogonalize.py

```diff
@@ -0,0 +1,165 @@
+"""Orthogonalization of vectors."""
+
+from typing import Callable, Iterable, Optional, Union
+
+import numpy as np
+
+from probnum import linops
+
+from ._inner_product import euclidean_inprod, euclidean_norm
+
+
+def gram_schmidt(
+    v: np.ndarray,
+    orthogonal_basis: Iterable[np.ndarray],
+    inprod: Optional[
+        Union[
+            np.ndarray,
+            linops.LinearOperator,
+            Callable[[np.ndarray, np.ndarray], np.ndarray],
+        ]
+    ] = None,
+    normalize: bool = False,
+) -> np.ndarray:
+    r"""Orthogonalize a vector with respect to an orthogonal basis and inner product.
+
+    Computes a vector :math:`v'` such that :math:`\langle v', b_i \rangle = 0` for
+    all basis vectors :math:`b_i \in B` in the orthogonal basis.
+
+    Parameters
+    ----------
+    v
+        Vector to orthogonalize.
+    orthogonal_basis
+        Orthogonal basis.
+    inprod
+        Inner product.
+    normalize
+        Normalize the output vector, s.t. :math:`\langle v', v' \rangle = 1`.
+
+    Returns
+    -------
+    v_orth
+        Orthogonalized vector.
+    """
+    if inprod is None:
+        inprod_fn = euclidean_inprod
+        norm_fn = euclidean_norm
+    elif isinstance(inprod, (np.ndarray, linops.LinearOperator)):
+        inprod_fn = lambda v, w: euclidean_inprod(v, w, A=inprod)
+        norm_fn = lambda v: euclidean_norm(v, A=inprod)
+    else:
+        inprod_fn = inprod
+        norm_fn = lambda v: np.sqrt(inprod_fn(v, v))
+
+    v_orth = v.copy()
+
+    for u in orthogonal_basis:
+        v_orth -= (inprod_fn(u, v) / inprod_fn(u, u)) * u
+
```

Comment:
# The loop in 57-58 can be vectorized
# if U is the orthonormal basis, then v_orth = v - UU^TAv
basis_norm = inprod_fn(orthogonal_basis,orthogonal_basis) # Compute norm of basis
orthonormal_basis = (basis_norm)**(-1/2)*orthogonal_basis.T).T # Divide basis vectors by norm
v_orth -= orthonormal_basis @ inprod_fn(orthonormal_basis,v)

Reply:
I initially tried to vectorize here, but quickly ran into usability issues. If we want to vectorize here easily, the user needs to specify an inner product, that behaves such that pairwise inner products are computed and not all pairs (like `np.inner` does). So here for now I would vote for ease of use over speed to avoid premature optimization.